### PR TITLE
Configurable Void Fog Height

### PR DESCRIPTION
**Pull request by ruvaldak**
This is still heavily WIP, but I figured I might as well open a PR now. Simply working on adding a slider to allow configuration of the height at which the void fog starts kicking in.

### CODE REVIEW

File: src/main/resources/assets/voidfog/lang/ru_ru.json [hunk 1]
```diff
@@ -2,11 +2,13 @@
...
+	"menu.voidfog.scale": "Scale With Difficulty",
 	"menu.voidfog.torches": "Спрятаться в освещенных местах",
 	"menu.voidfog.particles.min": "Без частиц",
 	"menu.voidfog.particles.max": "ВСЕ частицы",
 	"menu.voidfog.particles.default": "Частицы: По умолчанию",
 	"menu.voidfog.particles": "Частицы: %d",
+	"menu.voidfog.fogheight": "Fog Height: %d",
 	"menu.voidfog.bigboi": "Режим большого мальчика",
 	"menu.voidfog.bigboi.tooltip": "Делает игру еще более пугающей.\r\nВы большой парень?"
 }
```
Comment (Sollace) [inline src/main/resources/assets/voidfog/lang/ru_ru.json:16]: Please remove the english entries from this file. It should only contain the russian translations of the ones in en_us!

File: src/main/java/com/tamaized/voidfog/FogRenderer.java [hunk 2]
```diff
@@ -46,10 +46,26 @@ public void render(Camera camera,  FogType type, float viewDistance, boolean thi
...
+            //float end = VoidFog.config.maxFogHeight;
+            //float start = (VoidFog.config.maxFogHeight+15F);
+            //float entityDelta = 1-(((float)(entity.getY()-world.getBottomY()))-VoidFog.config.maxFogHeight)/15F;
+            RenderSystem.setShaderFogStart(MathHelper.lerp(1-(((float)(entity.getY()-world.getBottomY()))-VoidFog.config.maxFogHeight)/VoidFog.config.fadeStartHeight,RenderSystem.getShaderFogStart(),getFogStart(distance, type, world, thickFog)));
+            RenderSystem.setShaderFogEnd(MathHelper.lerp(1-(((float)(entity.getY()-world.getBottomY()))-VoidFog.config.maxFogHeight)/VoidFog.config.fadeStartHeight,RenderSystem.getShaderFogEnd(),getFogEnd(distance, type, world, thickFog)));
+        }
+        else {
+            RenderSystem.setShaderFogStart(RenderSystem.getShaderFogStart());
+            RenderSystem.setShaderFogEnd(RenderSystem.getShaderFogEnd());
+        }
```
Comment (Sollace) [inline src/main/java/com/tamaized/voidfog/FogRenderer.java:67]: Can you explain what all of these changes are doing?

Also please tidy up the formatting.

File: src/main/java/com/tamaized/voidfog/FogParticleSpawner.java [hunk 3]
```diff
@@ -22,7 +22,7 @@ public void update(World world, Entity entity, Voidable dimension) {
 
         Random rand = world.getRandom();
 
-        for (int pass = 0; pass < VoidFog.config.voidParticleDensity; pass++) {
+        for (int pass = 0; pass < VoidFog.config.voidParticleDensity; pass++) {
```
Comment (Sollace) [inline src/main/java/com/tamaized/voidfog/FogParticleSpawner.java]: Revert this change please

File: hs_err_pid71772.log [hunk 4]
```diff
@@ -0,0 +1,1349 @@
+#
```
Comment (Sollace) [inline hs_err_pid71772.log]: Remove this file please

Comment (Sollace) [reply hs_err_pid71772.log]: Beware this also contains your real name.

**Review by Sollace — CHANGES_REQUESTED** [review]

File: hs_err_pid71772.log (hunk 4 above)
Comment (ruvaldak) [reply hs_err_pid71772.log]: Yoink, thanks for the heads up

File: src/main/java/com/tamaized/voidfog/FogRenderer.java (hunk 2 above)
Comment (ruvaldak) [reply src/main/java/com/tamaized/voidfog/FogRenderer.java:67]: This was my attempt to fix #11 

First if checks to see if the player is below the void fog height or if fancier fog is enabled. If so, render the denser fog as mentioned in #11.
First else if checks to see if the player is within VoidFog.config.fadeStartHeight blocks of the void fog height. If so, begin lerping fog from vanilla fog to denser fog based on player height. (default value is 15 blocks. Didn't make it a slider because I wasn't sure how to implement it in a way that makes sense. VoidFog.config.fadeStartHeight is additive to VoidFog.config.maxFogHeight, not an absolute position)
Catchall else simply renders vanilla fog.

File: src/main/resources/assets/voidfog/lang/ru_ru.json (hunk 1 above)
Comment (ruvaldak) [reply src/main/resources/assets/voidfog/lang/ru_ru.json:16]: My bad - put that in there to test something and forgot to fix it.

File: src/main/java/com/tamaized/voidfog/FogParticleSpawner.java (hunk 3 above)
Comment (ruvaldak) [reply src/main/java/com/tamaized/voidfog/FogParticleSpawner.java]: Reverted in [8d465f2](https://github.com/Sollace/Void-Fog/pull/13/commits/8d465f21f9257c1788b6d29917d657585876508b)

File: hs_err_pid71772.log (hunk 4 above)
Comment (ruvaldak) [reply hs_err_pid71772.log]: Deleted in [95168f9](https://github.com/Sollace/Void-Fog/pull/13/commits/95168f96f06bcdbbb1a3950638afe597f7af1a3f)

File: src/main/java/com/tamaized/voidfog/FogRenderer.java (hunk 2 above)
Comment (ruvaldak) [reply src/main/java/com/tamaized/voidfog/FogRenderer.java:67]: Tidied up formatting in [7172273](https://github.com/Sollace/Void-Fog/pull/13/commits/71722730d9f5249fb12f839af0cba18aadfe904c), [0fb2078](https://github.com/Sollace/Void-Fog/pull/13/commits/0fb20785cb9b52e2be07b3ae03fcbcc487abea7f), and [8d51fb5](https://github.com/Sollace/Void-Fog/pull/13/commits/8d51fb5b933e07e430c4f3df292aff3740ff2429). I renamed fadeStartHeight to fadeStartOffset for clarity.

entityDelta is converting entityAltitude into a percentage (0.0-1.0 = 0% - 100%) of the range of VoidFog.config.maxFogHeight to fadeStart (maxFogHeight+fadeStartOffset), which is then used in the MathHelper.lerp line to determine how to transition from vanilla fog to the denser void fog. Because I'm only using MathHelper.lerp when the player is within that range, there should be a relatively smooth transition as the player approaches the void fog height. I used a default of 15 for fadeStartOffset, just to start with. Haven't really played with different numbers to see what looks best. Offset can technically be set to 0 to "disable" the fade. I might be able to turn that into an option - I should also probably do some checks to prevent negative numbers on the offset.

Comment (ruvaldak) [reply src/main/java/com/tamaized/voidfog/FogRenderer.java:67]: Added a check in [8dfd691](https://github.com/Sollace/Void-Fog/pull/13/commits/8dfd691d932c69278a9e985af9a81b14722d0278)

File: src/main/resources/assets/voidfog/lang/ru_ru.json (hunk 1 above)
Comment (ruvaldak) [reply src/main/resources/assets/voidfog/lang/ru_ru.json:16]: Fixed in [e0240d0](https://github.com/Sollace/Void-Fog/pull/13/commits/e0240d023896046a37fd7e0bb7a041be095248b1)

**Comment by ruvaldak** [comment]
This PR now does the following:
Adds configurable fog height
Fixes #11 by adding a transition from vanilla fog to denser void fog.
Adds a toggle for always denser fog and difficulty scaling.

Particles are unchanged and don't adjust with the fog height setting.

**Comment by ruvaldak** [comment]
Changes:
Particles now adjust according to fog height setting
Re-renamed alwaysDenseFog to prettyFog, because semantics.
changed fadeStartOffset's usage to be subtractive from maxFogHeight, instead of additive

Should be more or less done. Lemme know if I missed anything in particular or if any changes still need to be done.

**Review by Sollace — APPROVED** [review]
Alright, looks good. There's still some formatting weirdness, but nothing I can't fix. I'll have to also test this a little before I put it into a release, but otherwise looks good to merge.

**Comment by ruvaldak** [comment]
> Alright, looks good. There's still some formatting weirdness, but nothing I can't fix. I'll have to also test this a little before I put it into a release, but otherwise looks good to merge.

Happy to hear. Looking forward to hearing about how your testing goes
